### ISSUE TITLE
suggestion: STATIC_REQUIRE for testing constexpr code

### ISSUE THREAD

**Issue by DanGroomMPG** [onqtam/doctest#250]
I find this very simple macro very useful for testing constexpr code using doctest:

//fails compilation if the check fails, reports at runtime if it succeeds
#define STATIC_REQUIRE(x) static_assert(x); REQUIRE(x)

**Comment by onqtam** [comment]
I think Catch2 has something like this. It's a bit weird to be that reporting on runtime is desired if it already passed at compile time... I'll leave this open for now.

**Comment by DanGroomMPG** [comment]
It is a bit weird, agreed. Still, it helps to make constexpr tests more closely fit the existing patterns for runtime tests. Colleagues who are unused to compile time tests can still see when they run the executable and see the familiar doctest report that *something* has been checked - which to most readers is more important than whether it was already known to pass at compile time, I think.

**Comment by onqtam** [comment]
I'd advise using the #define locally for now (perhaps make a proxy doctest header which includes the actual one and defines extra struff) and I might add this in the future (for all asserts).

**Comment by DanGroomMPG** [comment]
Thanks, yes a proxy header that sets up global doctest config as well as defining this macro is exactly what I'm doing now :)

**Comment by Lectem** [comment]
I think it might actually be more interesting to force the expression to be evaluated at compile time but report it at runtime, even if it fails.

That way one can run the tests while not failing a build. 

For that, we could use something along the lines of 

```
#define STATIC_REQUIRE(x) constexpr bool expr = (x); REQUIRE(expr);
```

This would force the expression to be computed at compile time but still be reported at runtime. 
One could even try to test both compile time and runtime in a single define by duplicating the check and providing a named argument that will be constexpr in one block and runtime in another.

**Comment by kunaltyagi** [comment]
Forcing runtime or compile time both cases have their use. Perhaps a `DOCTEST_CONFIG_RUNTIME_STATIC_REQUIRE` like Catch2 too.

**Comment by s9w** [comment]
I'd like to +1 this. `constexpr` functions are somewhat fickle in that there are a lot of circumstances where a function compiles as `constexpr` but can't really be called at compile time due to the formulation of the standard in that regard. That leads to a common practice where a `static_assert()` is being written to enforce at least one compile time call.

However `static_assert()` is a crude tool. It can only contain a statement, and forcing variable definitions etc into a lambda is awkward. Combining all this into a constexpr doctest would be awesome. Because these tests should absoutely "count" towards the test/fail count during runtime.

**Comment by UnePierre** [comment]
> ```
> #define STATIC_REQUIRE(x) constexpr bool expr = (x); REQUIRE(expr);
> ```
> 
> This would force the expression to be computed at compile time [...]

I don't think this actually _forces_ the compiler into compile-time evaluation. It has to check compile-time-computability, but IMO it has the freedom to "delay" the actual evaluation until runtime in that construct, or remove it completely when the value is not needed. I haven't checked with the C++ (yet), though.

I would aim at putting the test result into a template <bool> parameter of a class that self-registers its result in some reportable list (or two separate lists for passed/failed). The pretty-printing could be done in lambda functions to avoid bloating the type room by passing stringyfied expressions via template arguments (which is possible, but ugly).

Oh, sorry: the self-registering isn't so easy in `constexpr`, either. I did a sketch of a linked list a while ago ...
It might be much more simple, if the test-result-container-instance needs not be constexpr.
But in that case, STATIC_REQUIRE would not be usable out-of-testcase, thus using it inside constexpr implementation functions wouldn't work.

That is BTW my intended use for a STATIC_REQUIRE: writing optimizable constexpr functions with defensive code, that produces nice output when tested. In functional safety, we do need to probe code with out-of-bounds values. If it were possible for "normal" REQUIRE to exist in constexpr context, I would be very happy!

**Comment by s9w** [comment]
Just as a practical hint for readers: I tried my hands on this locally, and the define as above doesn't work due to macro reasons. While my macro skills are rudimentary, this one actually works:

```c++
#define STATIC_CHECK(...) static_assert(__VA_ARGS__); CHECK(__VA_ARGS__);
```
Does a compile time static_assert, and runs it at runtime, too.

**Comment by UnePierre** [comment]
> ```c++
> #define STATIC_CHECK_FALSE(...) static_assert(!(__VA_ARGS__)); CHECK_FALSE(__VA_ARGS__);
> ```
Please always put parentheses around macro arguments during evaluation, or expected hell rains loose (try `STATIC_CHECK_FALSE(a && b);` for example).
